### ISSUE TITLE
Config for Synology - and where to place config.ovpn?

### ISSUE THREAD

**Issue by Ltek** [Kabe0/deluge-windscribe#8]
I'm working to set this up on a Synology NAS. I have it running but not sure its 100% config'd properly.

A few questions.

1... where do I place the config.ovpn file?

2... when I use the Windscribe Config generator, it only allows me to select 1 location, not a region AND require to select a specific port from a list (cannot input the ports I have set in Docker). 
-- Will I need to open a port in Docker to allow for Windscribe?
-- How do we create a file for use that allows it to select the 'best' as ypu stated was the default in the container setup?

3... Do I need the below commands or do they run automatically when I start the container? I use Synology's Docker manager UI - not scripts or compose
- if they need to be added separately, how do I put them in? As Environment variables or something else?

 --cap-add=NET_ADMIN \
 --device /dev/net/tun:/dev/net/tun \

thanks for helping out a noobie!

**Comment by Ltek** [comment]
Is this Image no longer maintained?

**Comment by Kabe0** [comment]
Sorry, Have been crazy busy. 

So you don't use a config.ovpn file with this setup. This uses windscribe-cli to handle the network interfacing. So you only to make a config file with the username/pass of your windscribe login, or define it as environment variables to the docker container as described in the example.

You do need to have the --device assigned, otherwise the docker container will not have access to the proper network driver to make the correct tunnel settings. --cap-add=NET_ADMIN gives the docker container access to modify the /net/tun as by default docker are not authorised to make network modifications.

I have used Synology NAS's in the past. What are you using to configure the docker container? I should be albe to respond faster now.

**Comment by Ltek** [comment]
@Kabe0 
thx for the info. I'm really new to this. I installed Portainer - def over my head for now.

Can you help me build the proper Compose file?

I've see people use some settings, not sure if I need to?

**Comment by Kabe0** [comment]
This is my actual docker-compose file I have had running for the past 6 months
```
version: "3"
services:
  deluge-windscribe:
    restart: unless-stopped
    image: kabe0/deluge-windscribe:latest
    volumes:
      - ./downloads:/downloads
      - ./DelugeConfig:/config
    ports:
      - 8112:8112
      - 8989:8989
      - 58846:58846
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun
```

A ./ represents the current directory. Inside the current directory, under DelugeConfig I have a config file called ```auth.conf``` with the three lines I mentioned above. The folde permissions for DelugeConfig are set to my user. You can find your user ID by typing
```id``` in your main system which will print out the uid and gid values. Make sure they match your variables on the container.

**Comment by Kabe0** [comment]
If you want send me the logs you get when you run it.

**Comment by Ltek** [comment]
I'm trying to run docker-compose against the YML (running as root) and all I get is... "Can't find a suitable configuration file in this directory or any parent. Are you in the right directory"

**Comment by Kabe0** [comment]
Docker is probably one of the easiest virtualization options to come around in a long time. It does have a little bit of a barrier. If you push through it though I think you will find it's extremely useful.

So the issue your running into is those ports are already consumed. So my guess is you have a container already running. 

Run

```docker ps``` to view all the active docker services. If you see one, look for it's unique instance name and end it by running

```docker stop <instance_name>```

Just to clarify, the ./media and ./docker folders are inside your folder with the docker-compose.yml right? It should look like this...
```
docker-compose.yml
media
     downloads
docker
     deluge-windscribe
         auth.conf
```

**Comment by Kabe0** [comment]
If you are running something on 8989 and don't want to change it, you can always rebind it to a different port by just specifying the alternative port on the right side of the ports defined, like

`18989:8989`

**Comment by Ltek** [comment]
just realized 8989 is the default port for Sonarr, I removed it and compose built the container. 
I cannot access it on port 8112 like I can with my non-VPN Container version of Deluge?

Does the error I got below matter?

```
root@NAS:~# docker-compose -f /volume1/docker/deluge-windscribe/docker-compose.yml up
Creating network "deluge-windscribe_default" with the default driver
Creating deluge-windscribe_deluge-windscribe_1 ... done
Attaching to deluge-windscribe_deluge-windscribe_1
deluge-windscribe_1  | WARNING: no logs are available with the 'db' log driver
deluge-windscribe_deluge-windscribe_1 exited with code 1
```

**Comment by Kabe0** [comment]
Just checking, your using the username/pass for your windscribe account not the openvpn right?

**Comment by Ltek** [comment]
Can I add DNS as an Environment variable, or do I need to rebuilt the container?
-- not sure why DNS would help me access it locally?

Yes, Username & password are correct

Wondering if I need this, I've seen this in other tutorials... command: "-r 192.168.2.0/24" # route local network traffic
how do I put that into a Compose YML?

**Comment by Kabe0** [comment]
So for some reason windscribe is not presenting a login request. So what would make sense to do now is test the container without any of the other pieces to see why it's not continuing. Run the following command anywhere to see if you can manually login to windscribe
```
docker run -it -p 8112:8112 -it --cap-add=NET_ADMIN \
 --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest /bin/bash
```
This command should launch you into the terminal for the docker container. In here we can run the windscribe-cli commands manually and see what is happening.

First, let's see if you have access to the web DNS. You can call ping on something like google. If it resolves properly you at least have internet in docker

```ping google.com```

Next, try running the following command to initiate windscribe-cli

```windscribe start```

This should print the following lines
```Starting windscribe ... OK```

If that happens, continue to the next step

```
windscribe login
```
Then, enter your username and password as you defined in the config file.

Let me know if that works, as right now, that's where the python script is failing. It's not recieving the message to put in the username and pass for windscribe.

**Comment by Ltek** [comment]
good news ... ALL of that worked perfectly. shows 'Logged In'

**Comment by Kabe0** [comment]
Oh we did not turn that part on yet, what this test was to see if you could get past that error message.

Now try running ```windscribe connect``` and see if you can get it to connect to one of the vpn locations. If so, we can now run the final test below and try running the container without any volumes. If this works, then at least we know the issue is not the container itself.

Remove that container, and then run this command (putting in your username and password) and let's see if it works.
```
docker run -e VPN_USERNAME=<username> \
 -e VPN_PASSWORD=<pass> \ -p 8112:8112 -it --cap-add=NET_ADMIN \
 --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest
```

If you have any dollar symbols in your password, you will need to escape them with a back slash like, ```VPN_PASSWORD=my\$Secret```

The script being run in the docker container is ```/usr/bin/python3 -u /usr/bin/init.py``` so if you want to /bin/bash again, that's the script being executed to start windscribe and docker.

If this does not work, then there is a bug that has appeared due to something on your system. Most likely something is causing the init.py file to fail to find the proper

**Comment by Ltek** [comment]
container connected...
```
root@ec275d9ffd8f:/# windscribe connect
Connecting to US East Washington DC Precedent (UDP:443)
Firewall Enabled
/etc/resolv.conf is not a symlink, this may break DNS
Connected to US East Washington DC Precedent
Your IP changed from 24.128.114.83 to 104.168.34.152
root@ec275d9ffd8f:/# 
```
when building the new container...

```
add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest
Initializing Container
Using VPN_USERNAME and VPN_PASSWORD to login.
Starting windscribe ... OK
Initializing Deluge
Configuring firewall settings
usermod: no changes
Making config directory.
Deluged Init
```
... the CLU hung here, did not give me back a prompt.
the Container is running and I can connect to the Web UI and add a torrent. BUT the DL doesnt start, Deluge shows "Error: Connection timed out" which I suspect is that there is no internet connection / VPN is not working?

**Comment by Kabe0** [comment]
So the CLU is not hanging, it's running as intended. We have not run the app in Daemon mode so it's printing out all the logs as is. To run it as a daemon just add -d at the end like
```
docker run -e VPN_USERNAME=<username> -e VPN_PASSWORD=<pass> -p 8112:8112 -it --cap-add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest -d
```
Does your deluge instance have an IP address on the bottom right?
![image](https://user-images.githubusercontent.com/6827834/104850601-053f5800-58be-11eb-86ad-cc28bbd885ba.png)


Anyways, so you got to the point of getting deluge to actually run which is good. That means we are closer to resolving whatever issue is going on. It looks like your config folder has an issue, so it's preventing deluge from running. Can you confirm that the use permissions for that folder are 1000:1000?

**Comment by Ltek** [comment]
All folders are set to an acct in the admin group... 1027 & 101
...on my other containers I set Enviro variables PUID & PGID to those.

I dont know how to set the folder perms to 1000:1000

**Comment by Kabe0** [comment]
oh the 1000:1000 is not mandatory, but based on what is happening, it looks like either a file or folder inside /config is set wrong. You can set your GID UID as follows
```
docker run -e VPN_USERNAME=<username> -e VPN_PASSWORD=<pass> -e DEL_UID=1027 -e DEL_GID=101  -p 8112:8112 -it --cap-add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest -d
```

I am just curious if you go inside your config folder, if you find that any of the file permissions are different than that of your main folder permissions?

**Comment by Ltek** [comment]
found this...

```
  File: ‘/volume1/docker/deluge-windscribe’
  Size: 68              Blocks: 0          IO Block: 4096   directory
Device: 2eh/46d Inode: 5556        Links: 1
Access: (0755/drwxr-xr-x)  Uid: ( 1000/ UNKNOWN)   Gid: ( 1000/ UNKNOWN)
Access: 2021-01-17 08:56:24.334448508 -0700
Modify: 2021-01-17 08:56:16.948359516 -0700
Change: 2021-01-17 08:56:17.214362712 -0700
 Birth: -
```

when I run your new command...
`docker: Error response from daemon: OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"-d\": executable file not found in $PATH": unknown.`

**Comment by Kabe0** [comment]
oh whoops, the -d is at the wrong spot
```
docker run -d -e VPN_USERNAME=<username> -e VPN_PASSWORD=<pass> -e DEL_UID=1027 -e DEL_GID=101 -p 8112:8112 -it --cap-add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-openvpntest kabe0/deluge-windscribe:latest

```

**Comment by Ltek** [comment]
Stops Unexpectedly after about 1 or 2 mins... Web UI never able to pull up

**Comment by Kabe0** [comment]
Ok, leave the -d off for now until the error is sorted out

if you run ``ll`` in your /config directory, what does it look like?

**Comment by Ltek** [comment]
Using your command line, it never creates a .config folder
If I add ... -v /volume1/docker/deluge-windscribe:/config  ... then it creates .config folder, which 'administrators' group has Read/Write -- same as the parent folders, including 'docker'

with, or without the volume statement, we get this...
```
Using VPN_USERNAME and VPN_PASSWORD to login.
Starting windscribe ... OK
Initializing Deluge
Configuring firewall settings
groupmod: GID '101' already exists
Making config directory.
Deluged Init
Traceback (most recent call last):
  File "/usr/bin/deluged", line 11, in <module>
    load_entry_point('deluge===2.0.3-2-201906121747-ubuntu18.04.1', 'console_scripts', 'deluged')()
  File "/usr/lib/python3/dist-packages/deluge/core/daemon_entry.py", line 87, in start_daemon
    options = parser.parse_args()
  File "/usr/lib/python3/dist-packages/deluge/argparserbase.py", line 250, in parse_args
    return self._handle_ui_options(options)
  File "/usr/lib/python3/dist-packages/deluge/argparserbase.py", line 343, in _handle_ui_options
    os.setuid(options.group)
PermissionError: [Errno 1] Operation not permitted
```

**Comment by Kabe0** [comment]
PermissionError: [Errno 1] Operation not permitted

that's the problem

**Comment by Ltek** [comment]
the Deluge container (using linuxserver image) runs perfectly fine... just doesnt have VPN built-in

**Comment by Ltek** [comment]
your command line without the volume statement, never creates the folder.

**Comment by Kabe0** [comment]
Volumes are part of docker volumes. If you don't bind to a real physical folder, it will store it inside dockers own volumes instead.

Did you use this config folder with linuxserver as well as my container? That might cause a problem. I don't know if their config is mounted to the home folder like mine is. Can you mount it to a new config folder instead?

The folder structure should look like this if it's structured properly
```
/volume1/docker/deluge-windscribe
  .config
    deluge
        core.conf
```

**Comment by Ltek** [comment]
if I add the volume statement, I still get PermissionError BUT it does create BOTH the .config folder and 'deluge' folder inside the .config folder + core.conf

**Comment by Kabe0** [comment]
one sec I think i got the bug to reproduce.

**Comment by Kabe0** [comment]
Ok so I want you to update your image. Run ```docker image pull kabe0/deluge-windscribe:latest```
Let me know if it runs now.

**Comment by Ltek** [comment]
updating it now... yes, Intel.

**Comment by Ltek** [comment]
its running and I can get to the WebUI... oddly, when I try to put in a folder in AutoAdd, I get ... "
You must have read and write access to watch folder."
```
20:01:17 [ERROR   ][deluge.ui.web.json_api     :185 ] [Failure instance: Traceback (failure with no frames): <class 'deluge.error.WrappedException'>: You must have read and write access to watch folder.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/deluge/core/rpcserver.py", line 326, in dispatch
    ret = self.factory.methods[method](*args, **kwargs)
  File "/usr/lib/python3/dist-packages/deluge/plugins/AutoAdd-1.8.egg/deluge_autoadd/core.py", line 446, in add
    'You must have read and write access to watch folder.',
  File "/usr/lib/python3/dist-packages/deluge/plugins/AutoAdd-1.8.egg/deluge_autoadd/core.py", line 79, in check_input
    raise Exception(message)
Exception: You must have read and write access to watch folder.
```

**Comment by Kabe0** [comment]
ok i'll have to look at this tonight. There is something happening. Not sure what you setup as autoadd unless your talking about a plugin?

**Comment by Ltek** [comment]
AND there's more...

I rebuilt the container:
removed command line volume... /media/downloads:/downloads mapping. 
added command line volume... /media:/media

then went into AutoAdd and tried to set Watch Folder to /media/downloads  ... "Path does not exist"

**Comment by Kabe0** [comment]
The issue is related to the group ID. I initially had a -u and -G flag set on the daemon for deluge, but that was throwing the error. I think that might be why your file issue is happening

Turning that off was not the real fix. This looks easy enough to solve. Just have to finish some actual work first before I look at this haha.

**Comment by Ltek** [comment]
AutoAdd plugin is built-in. And seems to be smart (vs Deluge being dumb) since it actually checks to ensure you get access to the folders you config it with.

I rebuilt the container:
removed command line volume... /media/downloads:/downloads mapping. 
added command line volume... /media:/media

then went into AutoAdd and tried to set Watch Folder to /media/downloads  ... "Path does not exist"

**Comment by Kabe0** [comment]
Oh your path is going to be /downloads not /media/downloads. Whatever you want to setup you have to link to the location you define inside the docker container. Then that should work.

I am right now testing the problem we had with the file permissions. Will let you know probably in the hour when it's working.

**Comment by Kabe0** [comment]
Ok got somewhere. Ran a test of user 1001:1002 as my sample.

Got a torrent downloading *ubuntu and managed to have it download to the downloads folder without a problem. Apologies for all the back and fourth. Most questions I get are usually easily solved so this is the first real bug that has come up.

---
The symptom was that you cannot run both the user and group calls with the deluged service. Now, this might not be an actual issue in pure daemon mode, but that would require configuring the logs differently and personally I like my containers to run a bit lighter than that if possible. The group is already the primary for the user anyways so there is really no reason to assign the group explicitly anyways to the container.

**Comment by Kabe0** [comment]
Make sure by the way for your torrent watcher permissions match that of the uid:gid, or you can change the permissions of the folder so that it at least has read access. I have just tested the watcher to make sure it will work for you
![image](https://user-images.githubusercontent.com/6827834/104872019-ac51dd00-591a-11eb-9412-c4c9cb359d0a.png)

**Comment by Ltek** [comment]
I tried it again and it ran and the Web UI came up... but only the first time. After I shut it down, the Web UI never came back up. And when I start it manually again, it crashes.

I need to map a volume to the /volume1/media folder. -- which works perfectly on the linuxserver non-VPN Deluge container using the same UID/GID. That container they ask use to set PGID & PUID in Enviro Variables.  I also have containers from linuxserver for Radarr, Sonarr, Jackett -- and Portainer running 24x7 using the same folder structure / mapped volumes.

I'm using this command to build yours...

`docker run -e VPN_USERNAME=<USER> -e VPN_PASSWORD=<PW> -v /volume1/docker/deluge-windscribe:/config -v /volume1/media:/media -e DEL_UID=1027 -e DEL_GID=100 -p 58846:58846 -p 8112:8112 -it --cap-add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-windscribe kabe0/deluge-windscribe:latest`

I'm at a loss why the linuxserver version worked perfectly first shot and yours is not. I'm thinking it might just be better to figure out how to run a container that is just a VPN and route my existing linuxserver Deluge container through it. I've spent like 3 hours trying to get yours working and its not.

thanks for the time.

**Comment by Kabe0** [comment]
So I see the issue your having here with start/stop. I typically use docker-compose so did not notice this problem. Adding the dns flag resolved the issue for me, but will need to look at it at some point in the future. It's an issue with the VPN interfering with the resolve.conf.

docker run -e VPN_USERNAME=<USER> -e VPN_PASSWORD=<PW> -v /volume1/docker/deluge-windscribe:/config -v /volume1/media:/media -e DEL_UID=1027 --dns 8.8.8.8 -e DEL_GID=100 -p 58846:58846 -p 8112:8112 -it --cap-add=NET_ADMIN --device /dev/net/tun:/dev/net/tun --name deluge-windscribe kabe0/deluge-windscribe:latest

**Comment by Ltek** [comment]
Added --dns seems to have fixed startup

Still cant access the volumes. It sure looks like you have hard coded /downloads since when I open Deluge for the first time, it shows /downloads filled in for the settings?

Tried these mappings, not at same time... UID is THE Admin!

error in Deluge... "You must have read and write access to watch folder."

/volume1/media/downloads --> /downloads     
/volume1/media  --> /media
/volume1/media  --> /store     **tried also since you said 'media' cant be used

**Comment by Kabe0** [comment]
Downloads is hard-coded as a lot of users were having problems with the first-time startup, so I created a template that also restricts the access to the VPN only to ensure there is no way packets are accidentally sent if the VPN for whatever reason disconnects, however it's just a deluge config, so whatever you change it to persists onwards as long as there is a deluge folder.

Ah I get it not sure why it took me that long to see it. Your using the preserved group users = 100. The linuxserver image is a custom image so they avoid group conflicts. I am using Ubuntu as my base so if a group name is reserved. The solution is kind of silly, but I just had to add a few base groups to the deluge user and it's good to go. I pushed users into the deluge user as it does make sense to have that user part of that group pool.

The latest image pushed has the user added to the group. A hotfix on an older imagewould be just running this command
```usermod -a -G users deluge``` though it's not a great solution to run every restart.

![image](https://user-images.githubusercontent.com/6827834/104952787-2c1b8e00-5993-11eb-8154-5b196ed53e8a.png)

For future reference, if your ever working with docker images, typically ID's greater than 999 as anything below that is system reserved. Some of the big image vendors will just tell you to add your requirements to a custom Dockerfile :P
